src/compare-tree/key-path.js

```diff
@@ -0,0 +1,40 @@
+const KEY_DELIMETER  = '.'
+const PATH_DELIMETER = ','
+
+import {
+  castPath
+} from '../utils'
+
+function byKeyLength (a, b) {
+  return a.length > b.length ? 1 : -1
+}
+
+export function getKeyString (value) {
+  return value.join(KEY_DELIMETER)
+}
+
+export function getKeyPaths (value) {
+  if (value == null) {
+    return []
+  }
+
+  let paths = value
+
+  if (Array.isArray(value) === false) {
+    paths = `${paths}`.split(PATH_DELIMETER)
+  }
+
+  return paths.map(castPath)
+}
+
+export function getKeyStrings (array) {
+  let sorted = array.slice(0).sort(byKeyLength)
+  let key = ''
+
+  for (var i = 0, len = sorted.length; i < len; i++) {
+    key += i > 0 ? PATH_DELIMETER : ''
+    key += getKeyString(sorted[i])
+  }
+
+  return key
```

Comment:
i'm tempted to suggest pulling this out, makes it tougher to understand that the id for the query you generate is likely just the same thing that you've passed in to `on` to begin with. and it protects against a use case that i imagine would be quite rare, as well as not something i would really expect to be covered (turning on and off with the same key seems like a reasonable requirement).